lib/integrations/awsoidc/deployservice.go

```diff
@@ -0,0 +1,696 @@
+		TokenName: string(types.JoinMethodIAM) + "-token",
+		Method:    types.JoinMethodIAM,
+	}
+
+	switch req.DeploymentMode {
+	case DatabaseServiceDeploymentMode:
+		teleportConfig.Databases.Service.EnabledFlag = "yes"
+		teleportConfig.Databases.ResourceMatchers = []config.ResourceMatcher{{
+			Labels: req.DatabaseResourceMatcherLabels,
+		}}
+
+	default:
+		return "", trace.Errorf("invalid deployment mode, supported modes: %v", DeploymentModes)
+	}
+
+	teleportConfigYAMLBytes, err := yaml.Marshal(teleportConfig)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// This Config is meant to be passed as argument to `teleport start`
+	// Eg, `teleport start --config-string <X>`
+	teleportConfigString := base64.StdEncoding.EncodeToString(teleportConfigYAMLBytes)
+
+	return teleportConfigString, nil
+}
+
+// upsertTask ensures a TaskDefinition with TaskName exists
+func upsertTask(ctx context.Context, clt DeployServiceClient, req DeployServiceRequest, configB64 string) (*ecsTypes.TaskDefinition, error) {
+	taskAgentContainerImage := fmt.Sprintf(teleportContainerImageFmt, teleport.Version)
+
+	taskDefOut, err := clt.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+		Family: req.TaskName,
+		RequiresCompatibilities: []ecsTypes.Compatibility{
+			ecsTypes.CompatibilityFargate,
+		},
+		Cpu:    &taskCPU,
+		Memory: &taskMem,
+
+		NetworkMode:      ecsTypes.NetworkModeAwsvpc,
+		TaskRoleArn:      &req.TaskRoleARN,
+		ExecutionRoleArn: &req.TaskRoleARN,
+		ContainerDefinitions: []ecsTypes.ContainerDefinition{{
+			Command: []string{
+				"start",
+				"--config-string",
+				configB64,
+			},
+			EntryPoint: []string{"teleport"},
+			Image:      &taskAgentContainerImage,
+			Name:       &taskAgentContainerName,
+			LogConfiguration: &ecsTypes.LogConfiguration{
+				LogDriver: ecsTypes.LogDriverAwslogs,
+				Options: map[string]string{
+					"awslogs-group":         "ecs-" + *req.ClusterName,
+					"awslogs-region":        req.Region,
+					"awslogs-create-group":  "true",
+					"awslogs-stream-prefix": *req.ServiceName + "/" + *req.TaskName,
+				},
+			},
+		}},
+		Tags: req.ResourceCreationTags.ForECS(),
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return taskDefOut.TaskDefinition, nil
+}
+
+// upsertCluster creates the cluster if it doesn't exist.
+// It will update the cluster if it doesn't have the required capacity provider (FARGATE)
+// It will re-create if its status is INACTIVE.
+// If the cluster status is not ACTIVE, an error is returned.
+// The cluster is returned.
+func upsertCluster(ctx context.Context, clt DeployServiceClient, req DeployServiceRequest) (*ecsTypes.Cluster, error) {
+	describeClustersResponse, err := clt.DescribeClusters(ctx, &ecs.DescribeClustersInput{
+		Clusters: []string{*req.ClusterName},
+		Include: []ecsTypes.ClusterField{
+			ecsTypes.ClusterFieldTags,
+		},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if clusterMustBeCreated(describeClustersResponse.Clusters) {
```

Comment:
What's the usecase for installing into existing cluster instead of provisioning a new one?

Reply:
It only installs into existing cluster **if the cluster has the same name and it has the same ownership tags**.
Having the same name, means that it belongs to the same cluster.
Having the ownership tags, means it was created by this cluster and integration.

This way we have a single ECS Cluster per Teleport Cluster/Integration.

The alternative would be to create a new ECS Cluster every time, which seems not ideal from the management point of view (we will eventually have APIs to manage the Cluster' services).

Do you suggest another approach?